src/System.Private.ServiceModel/tests/Scenarios/Extensibility/WebSockets/WebSocketTests.cs

```diff
@@ -0,0 +1,234 @@
+            client = channelFactory.CreateChannel();
+
+            // *** EXECUTE *** \\
+            using (Stream stream = client.DownloadStream())
+            {
+                int readResult;
+                // Read from the stream, 1000 bytes at a time.
+                byte[] buffer = new byte[1000];
+
+                do
+                {
+                    readResult = stream.Read(buffer, 0, buffer.Length);
+                }
+                while (readResult != 0);
+            }
+
+            uploadStream = new FlowControlledStream();
+            uploadStream.ReadThrottle = TimeSpan.FromMilliseconds(500);
+            uploadStream.StreamDuration = TimeSpan.FromSeconds(1);
+            client.UploadStream(uploadStream);
+
+            // *** VALIDATE *** \\
+            foreach (string serverLogItem in client.GetLog())
+            {
+                Assert.True(serverLogItem != ScenarioTestHelpers.RemoteEndpointMessagePropertyFailure, ScenarioTestHelpers.RemoteEndpointMessagePropertyFailure);
+            }
+
+            // *** CLEANUP *** \\
+            ((ICommunicationObject)client).Close();
+            channelFactory.Close();
+        }
+        finally
+        {
+            // *** ENSURE CLEANUP *** \\
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)client, channelFactory);
+        }
+    }
+
+    [Fact]
+    [OuterLoop]
+    [ActiveIssue(424)]
+    public static void WebSocketHttpDuplexBinaryStreamed()
+    {
+        NetHttpBinding binding = null;
+        ClientReceiver clientReceiver = null;
+        InstanceContext context = null;
+        DuplexChannelFactory<IWSDuplexService> channelFactory = null;
+        IWSDuplexService client = null;
+        FlowControlledStream uploadStream = null;
+
+        try
+        {
+            // *** SETUP *** \\
+            binding = new NetHttpBinding();
+            binding.WebSocketSettings.TransportUsage = WebSocketTransportUsage.Always;
+            binding.MaxReceivedMessageSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize;
+            binding.MaxBufferSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize;
+            binding.TransferMode = TransferMode.Streamed;
+            binding.MessageEncoding = NetHttpMessageEncoding.Binary;
+
+            clientReceiver = new ClientReceiver();
+            context = new InstanceContext(clientReceiver);
+
+            channelFactory = new DuplexChannelFactory<IWSDuplexService>(context, binding, Endpoints.WebSocketHttpDuplexBinaryStreamed_Address);
+            client = channelFactory.CreateChannel();
+
+            // *** EXECUTE *** \\
+            using (Stream stream = client.DownloadStream())
+            {
+                int readResult;
+                // Read from the stream, 1000 bytes at a time.
+                byte[] buffer = new byte[1000];
+                do
+                {
+                    readResult = stream.Read(buffer, 0, buffer.Length);
+                }
```

Comment:
Question only: Can this synchronous Stream.Read block forever in any error conditions?  Just checking if we need to think about an async read with a timeout or something like that.